LibraBFT/Impl/Storage/DiemDB/DiemDB.agda

```diff
@@ -28,22 +47,86 @@ module getEpochEndingLedgerInfos where
 
 getEpochEndingLedgerInfos = getEpochEndingLedgerInfos.E
 
-module saveTransactions where
+-- TODO-2: hook this up with above
+-- Returns ledger infos for epoch changes starting with the given epoch.
+-- If there are less than `MAX_NUM_EPOCH_ENDING_LEDGER_INFO` results, it returns all of them.
+-- Otherwise the first `MAX_NUM_EPOCH_ENDING_LEDGER_INFO` results are returned
+-- and a flag is set to True to indicate there are more.
+getEpochEndingLedgerInfosX
+  : DiemDB → Epoch → Epoch
+  → Either ErrLog (List LedgerInfoWithSignatures × Bool)
+getEpochEndingLedgerInfosX self startEpoch endEpoch =
+  getEpochEndingLedgerInfosImpl self startEpoch endEpoch mAX_NUM_EPOCH_ENDING_LEDGER_INFO
+
+getEpochEndingLedgerInfosImpl self startEpoch endEpoch limit = do
+  if (not ⌊ (startEpoch ≤? endEpoch) ⌋)
+    then
+    (Left fakeErr {-here ["bad epoch range", lsE startEpoch, lsE endEpoch]-})
+    else pure unit
+  -- the latest epoch can be the same as the current epoch (in most cases), or
+  -- current_epoch + 1 (when the latest ledger_info carries next validator set)
+  latestEpoch ← getLatestLedgerInfo self >>= pure ∘ (_^∙  liwsLedgerInfo ∙ liNextBlockEpoch)
+  if (not ⌊ (endEpoch ≤? latestEpoch) ⌋)
+    then
+    (Left fakeErr) -- [ "unable to provide epoch change ledger info for still open epoch"
+                   -- , "asked upper bound", lsE endEpoch
+                   -- , "last sealed epoch", lsE (latestEpoch - 1) ])))
+                   -- -1 is OK because genesis LedgerInfo has .next_block_epoch() == 1
+    else pure unit
+
+  let (pagingEpoch , more) =
+        ECP-LBFT-OBM-Diff-2.e_DiemDB_getEpochEndingLedgerInfosImpl_limit startEpoch endEpoch limit
+  lis0    ← LedgerStore.getEpochEndingLedgerInfoIter (self ^∙ ddbLedgerStore) startEpoch pagingEpoch
+  let lis = LedgerStore.obmEELIICollect lis0
+  if length lis /= (pagingEpoch {-^∙ eEpoch-}) ∸ (startEpoch {-^∙ eEpoch-})
```

Comment:
I changed Haskell too.
I also left the old version (commented out) in both with a comment it was rewritten to avoid monus.
I also did some checking to ensure it is accurate.